apis/v1alpha1/gateway_types.go

```diff
@@ -717,38 +717,140 @@ type ListenerStatus struct {
 // field.
 type ListenerConditionType string
 
+// ListenerConditionReason defined the set of reasons that explain
+// why a particular Listener condition type has been raised.
+type ListenerConditionReason string
+
+const (
+	// ListenerConditionConflicted indicates that the controller
+	// was unable to resolve conflicting specification requirements
+	// for this Listener. If a Listener is conflicted, its network
+	// port should not be configured on any network elements.
+	//
+	// Possible reasons for this condition to be true are:
+	//
+	// * "HostnameConflict"
+	// * "ProtocolConflict"
+	// * "RouteConflict"
+	//
+	// Controllers may raise this condition with other reasons,
+	// but should prefer to use the reasons listed above to improve
+	// interoperability.
+	ListenerConditionConflicted ListenerConditionType = "Conflicted"
+
+	// ListenerReasonHostnameConflict is used when the Listener
+	// violates the Hostname match constraints for that allow
+	// collapsing Listeners. For example, this reason would be
+	// used when multiple Listeners on the same port use the
+	// "Any" hostname match type.
+	ListenerReasonHostnameConflict ListenerConditionReason = "HostnameConflict"
+
+	// ListenerReasonProtocolConflict is used when multiple
+	// Listeners are specified with the same Listener port number,
+	// but have conflicting protocol specifications.
+	ListenerReasonProtocolConflict ListenerConditionReason = "ProtocolConflict"
+
+	// ListenerReasonRouteConflict is used when the route
+	// resources selected for this Listener conflict with other
+	// specified properties of the Listener (e.g. Protocol).
+	// For example, a Listener that specifies "UDP" as the protocol
+	// but a route selector that resolves "TCPRoute" objects.
+	ListenerReasonRouteConflict ListenerConditionReason = "RouteConflict"
+)
+
+const (
+	// ListenerConditionDetached indicates that, even though
+	// the listener is syntactically and semantically valid, the
+	// controller is not able to configure it on the underlying
+	// Gateway infrastructure.
```

Comment:
What happens if the Gateway was already valid and something changes to arrive at this condition? This almost makes it sound like the Gateway would not be provisioned/updated at all.

Reply:
This is a similar problem to the conflicts. Say you have a Listener with protocol "TCP" and the operator changes the protocol top "FOU". The gateway may want to reject that change, but there's no way to show that the protocol is still "TCP" while also expressing that the "FOU" protocol change hasn't been applied. If the conditions are all positive, then the operator would expect that the "FOU" protocol is satisfactorily configured.

I suppose that a controller could say the listener is Detached/UnsupportedProtocol and still leave the "TCP" protocol in place. I'm not sure how I feel about that, since it's requiring Spec to vary from Status in a semi-permanent way.

Reply:
OK, that's a good clarification. This simply indicates that this specific listener would be deprovisioned in the event of invalid configuration here?